// @shopify/polaris/build/cjs/components/Modal/components/Dialog/Dialog.js
'use strict';

var React = require('react');
var reactTransitionGroup = require('react-transition-group');
var css = require('../../../../utilities/css.js');
var focus = require('../../../../utilities/focus.js');
var types = require('../../../../types.js');
var useTheme = require('../../../../utilities/use-theme.js');
var Dialog$1 = require('./Dialog.scss.js');
var context = require('../../../../utilities/frame/context.js');
var TrapFocus = require('../../../TrapFocus/TrapFocus.js');
var Text = require('../../../Text/Text.js');
var KeypressListener = require('../../../KeypressListener/KeypressListener.js');

function Dialog({
  instant,
  labelledBy,
  children,
  limitHeight,
  size,
  onClose,
  onExited,
  onEntered,
  setClosing,
  hasToasts,
  ...props
}) {
  const theme = useTheme.useTheme();
  const containerNode = React.useRef(null);
  const frameContext = React.useContext(context.FrameContext);
  let toastMessages;
  if (frameContext) {
    toastMessages = frameContext.toastMessages;
  }
  const classes = css.classNames(Dialog$1.default.Modal, size && Dialog$1.default[css.variationName('size', size)], limitHeight && Dialog$1.default.limitHeight);
  const TransitionChild = instant ? reactTransitionGroup.Transition : FadeUp;
  React.useEffect(() => {
    containerNode.current && !containerNode.current.contains(document.activeElement) && focus.focusFirstFocusableNode(containerNode.current);
  }, []);
  const handleKeyDown = () => {
    if (setClosing) {
      setClosing(true);
    }
  };
  const handleKeyUp = () => {
    if (setClosing) {
      setClosing(false);
    }
    onClose();
  };
  const ariaLiveAnnouncements = /*#__PURE__*/React.createElement("div", {
    "aria-live": "assertive"
  }, toastMessages ? toastMessages.map(toastMessage => /*#__PURE__*/React.createElement(Text.Text, {
    visuallyHidden: true,
    as: "p",
    key: toastMessage.id
  }, toastMessage.content)) : null);
  return /*#__PURE__*/React.createElement(TransitionChild, Object.assign({}, props, {
    nodeRef: containerNode,
    mountOnEnter: true,
    unmountOnExit: true,
    timeout: parseInt(theme.motion['motion-duration-200'], 10),
    onEntered: onEntered,
    onExited: onExited
  }), /*#__PURE__*/React.createElement("div", {
    className: Dialog$1.default.Container,
    "data-polaris-layer": true,
    "data-polaris-overlay": true,
    ref: containerNode
  }, /*#__PURE__*/React.createElement(TrapFocus.TrapFocus, null, /*#__PURE__*/React.createElement("div", {
    role: "dialog",
    "aria-modal": true,
    "aria-label": labelledBy,
    "aria-labelledby": labelledBy,
    tabIndex: -1,
    className: Dialog$1.default.Dialog
  }, /*#__PURE__*/React.createElement("div", {
    className: classes
  }, /*#__PURE__*/React.createElement(KeypressListener.KeypressListener, {
    keyCode: types.Key.Escape,
    keyEvent: "keydown",
    handler: handleKeyDown
  }), /*#__PURE__*/React.createElement(KeypressListener.KeypressListener, {
    keyCode: types.Key.Escape,
    handler: handleKeyUp
  }), children), ariaLiveAnnouncements))));
}
const fadeUpClasses = {
  appear: css.classNames(Dialog$1.default.animateFadeUp, Dialog$1.default.entering),
  appearActive: css.classNames(Dialog$1.default.animateFadeUp, Dialog$1.default.entered),
  enter: css.classNames(Dialog$1.default.animateFadeUp, Dialog$1.default.entering),
  enterActive: css.classNames(Dialog$1.default.animateFadeUp, Dialog$1.default.entered),
  exit: css.classNames(Dialog$1.default.animateFadeUp, Dialog$1.default.exiting),
  exitActive: css.classNames(Dialog$1.default.animateFadeUp, Dialog$1.default.exited)
};
function FadeUp({
  children,
  ...props
}) {
  return /*#__PURE__*/React.createElement(reactTransitionGroup.CSSTransition, Object.assign({}, props, {
    classNames: fadeUpClasses
  }), children);
}

exports.Dialog = Dialog;
